File: server/workflows.py
"""Declarative workflow definitions, read from YAML.

Three files describe the conversation. `entry` is the catalogue of domains a
first answer routes into. `operations` is the global list of things the
conversation can ask for — no domain may invent one. A domain file supplies
only the wording for those operations.

Control flow is deliberately absent here: which operation to ask about is
decided in `router`, from the graph itself. A new domain is therefore wording
and nothing else.
"""

from functools import cache
from pathlib import Path

import yaml
from pydantic import BaseModel, Field

WORKFLOW_DIR = Path(__file__).resolve().parent.parent / "workflows"


class WorkflowStep(BaseModel):
    """One prompt shown to the user, and what an answer may look like.

    `action` is the intent an answer is expected to become. A question asks for
    one thing, so naming it keeps the model filling known slots rather than
    choosing among every action there is."""

    id: str
    prompt: str
    choices: list[str] = Field(default_factory=list)
    hint: str = ""
    action: str = ""
    placeholder: str = ""


class Template(BaseModel):
    """Catalogue entry: one chip on the opening question, and the domain it
    opens. `about` is what the chip means, and what free text is scored against."""

    id: str
    chip: str
    about: str = ""


class Entry(BaseModel):
    """The opening question, and the domains a first answer routes into."""

    name: str = "entry"
    welcome: list[str] = Field(default_factory=list)
    placeholder: str = ""
    hint: str = ""
    templates: list[Template] = Field(default_factory=list)

    def chip(self, template_id: str) -> str:
        """Wording of one domain's chip, for naming what the user just chose."""
        return next((t.chip for t in self.templates if t.id == template_id), template_id)


class Operation(BaseModel):
    """One thing the conversation can ask for, and the intent an answer becomes.

    `when` names the condition under which there is anything to ask; `router`
    is what evaluates it."""

    id: str
    action: str
    chip: str = ""
    when: str = "always"


class Operations(BaseModel):
    """The global operation list, in the order they are preferred by default."""

    name: str = "operations"
    operations: list[Operation] = Field(default_factory=list)


class Wording(BaseModel):
    """What one domain says when it asks for one operation. `{label}` in either
    line is filled with the selected document's name."""

    prompt: str
    hint: str = ""
    choices: list[str] = Field(default_factory=list)


class Domain(BaseModel):
    """A domain's wording for the global operations.

    `lead` names the operation this domain prefers to open with — evidence
    comes before decomposition in research, for instance."""

    name: str
    lead: str = ""
    prompts: dict[str, Wording] = Field(default_factory=dict)

    def wording(self, operation: str, root: bool) -> Wording | None:
        """Wording for one operation, preferring the `_root` variant when the
        project itself is selected and there is no document to name."""
        keys = [f"{operation}_root", operation] if root else [operation]

        return next((self.prompts[key] for key in keys if key in self.prompts), None)


def _read(name: str, directory: Path) -> dict:
    """Parse one workflow file."""
    return yaml.safe_load((Path(directory) / f"{name}.yaml").read_text(encoding="utf-8"))


@cache
def entry(directory: Path = WORKFLOW_DIR) -> Entry:
    """The domain catalogue."""
    return Entry.model_validate(_read("entry", directory))


@cache
def operations(directory: Path = WORKFLOW_DIR) -> list[Operation]:
    """The global operation list."""
    return Operations.model_validate(_read("operations", directory)).operations


@cache
def domain(name: str, directory: Path = WORKFLOW_DIR) -> Domain:
    """One domain's wording, falling back to the catch-all if it has none."""
    try:
        return Domain.model_validate(_read(name, directory))
    except FileNotFoundError:
        return Domain.model_validate(_read("freeform", directory))

File: web/src/core/workflows.ts
/** Workflow definitions, read from YAML at build time.
 *
 *  Three kinds of file. `entry` is the catalogue of domains a first answer
 *  routes into. `operations` is the global list of things the conversation can
 *  ask for — no domain may invent one. A domain file supplies wording and
 *  vocabulary, and nothing else.
 *
 *  Control flow is deliberately absent: which operation to ask about is
 *  decided in `router`, from the graph itself. A new domain is wording only. */

import entryFile from "../../../workflows/entry.yaml";
import operationsFile from "../../../workflows/operations.yaml";

export type Template = {
  id: string;
  chip: string;
  about: string;
  /** Short phrases naming what someone might be making. Each is scored
   *  separately, so they are phrases rather than keywords. */
  tags: string[];
};

export type Entry = {
  welcome: string[];
  placeholder: string;
  hint: string;
  templates: Template[];
};

export type Operation = {
  id: string;
  chip: string;
  /** Condition under which there is anything to ask; see router.eligible. */
  when: string;
};

/** What a domain calls things, used for chips and the object explorer. */
export type Terms = { group: string; node: string; relation: string };

export type Wording = {
  /** One is chosen at random per asking, so the loop does not read as a form. */
  prompt: string[];
  hint: string;
  choices: string[];
};

export type Domain = {
  name: string;
  lead: string;
  terms: Terms;
  /** Relation kinds a new project in this domain starts with. */
  relations: string[];
  prompts: Record<string, Wording>;
};

const GENERIC: Terms = { group: "Group", node: "Object", relation: "Relation" };

/** Every domain file in the folder, minus the two that are not domains. */
const files = import.meta.glob("../../../workflows/*.yaml", {
  eager: true,
  import: "default",
}) as Record<string, any>;

function lines(value: unknown): string[] {
  if (Array.isArray(value)) return value.map(String);

  return value ? [String(value)] : [];
}

function wording(raw: any): Wording {
  return {
    prompt: lines(raw?.prompt),
    hint: raw?.hint ? String(raw.hint) : "",
    choices: Array.isArray(raw?.choices) ? raw.choices.map(String) : [],
  };
}

function domain(raw: any): Domain {
  const prompts: Record<string, Wording> = {};
  for (const [key, value] of Object.entries(raw?.prompts ?? {})) {
    prompts[key] = wording(value);
  }

  return {
    name: String(raw?.name ?? "freeform"),
    lead: String(raw?.lead ?? ""),
    terms: { ...GENERIC, ...(raw?.terms ?? {}) },
    relations: lines(raw?.relations),
    prompts,
  };
}

const DOMAINS: Record<string, Domain> = Object.fromEntries(
  Object.entries(files)
    .filter(([path]) => !/(entry|operations)\.yaml$/.test(path))
    .map(([, raw]) => [String(raw.name), domain(raw)]),
);

export const entry: Entry = {
  welcome: lines(entryFile.welcome),
  placeholder: String(entryFile.placeholder ?? ""),
  hint: String(entryFile.hint ?? ""),
  templates: (entryFile.templates ?? []).map((t: any) => ({
    id: String(t.id),
    chip: String(t.chip),
    about: String(t.about ?? ""),
    tags: lines(t.tags),
  })),
};

export const operations: Operation[] = (operationsFile.operations ?? []).map((o: any) => ({
  id: String(o.id),
  chip: String(o.chip ?? o.id),
  when: String(o.when ?? "always"),
}));

/** One domain's wording, falling back to the catch-all if it has no file. */
export function getDomain(name: string): Domain {
  return DOMAINS[name] ?? DOMAINS.freeform;
}

/** Wording for an operation, preferring the `_root` variant when the project
 *  itself is selected and there is no object to name. */
export function getWording(domain: Domain, operation: string, root: boolean): Wording | null {
  const keys = root ? [`${operation}_root`, operation] : [operation];

  return keys.map((key) => domain.prompts[key]).find(Boolean) ?? null;
}

/** The chip wording for a template id, for naming what the user just chose. */
export function chipFor(id: string): string {
  return entry.templates.find((t) => t.id === id)?.chip ?? id;
}
